packager/file/udp_file.cc

```diff
@@ -43,6 +47,85 @@ bool IsIpv4MulticastAddress(const struct in_addr& addr) {
   return (ntohl(addr.s_addr) & 0xf0000000) == 0xe0000000;
 }
 
+bool IsIpv6MulticastAddress(const struct in6_addr& addr) {
+  if (addr.s6_addr[0] == 0xff) {
+    return true;
+  }
+
+  return false;
+}
+
+template <typename T>
+bool SpecificMalticast(int ai_family,
+                       int level,
+                       void *imr_interface,
+                       void *imr_sourceaddr,
+                       SOCKET socket,
+                       T& source_multicast_group,
+                       std::unique_ptr<UdpOptions>&& options) {
+  if (inet_pton(ai_family,
+                options->interface_address().c_str(),
+                imr_interface) != 1) {
+    LOG(ERROR) << "Malformed IP interface address "
+               << options->interface_address();
+    return false;
+  }
+  if (inet_pton(ai_family,
+                options->source_address().c_str(),
+                imr_sourceaddr) != 1) {
+    LOG(ERROR) << "Malformed IP source specific multicast address "
+               << options->source_address();
+    return false;
+  }
+  if (setsockopt(socket,
+                 level,
+                 MCAST_JOIN_GROUP,
```

Comment:
Do you mean MCAST_JOIN_SOURCE_GROUP here?

Also, the argument should be struct GROUP_SOURCE_REQ instead of ip_mreq_source / ipv6_mreq.